schemactl/schema.py

```diff
@@ -0,0 +1,227 @@
+from enum import Enum
+import glob
+from os import path
+import os
+import re
+import textwrap
+from typing import Optional
+import yaml
+
+from functools import cache
+
+from pydantic import UUID4, BaseModel, ConfigDict, ValidationError, field_validator, model_validator
+
+def r2mm_dir() -> str:
+    dir_path = os.environ.get("R2MM_DIR", default="../games/r2modmanPlus/")
+    if os.path.isdir(dir_path):
+        return dir_path
+
+    raise Exception(textwrap.dedent(f"""
+        The r2modman directory at `{dir_path}` does not exist. You can fix this by either:
+        - Ensuring that this path exists and points to a valid local r2modman repository.
+        - Set a different r2modman directory path via the `R2MM_DIR` environment variable.
+    """))
+
+# Get a cached deserialized list of all ecosystem entries found within ../games/data.
+@cache
+def get_schema_entries(dir: str, skip_validation=False) -> list["SchemaEntry"] | list["PartialSchemaEntry"]:
+    schema_files = glob.glob(dir + "/*.yml")
+    yml_entries = map(lambda file: yaml.load(open(file), Loader=yaml.Loader), schema_files)
+
+    entries = []
+    for yml_entry in yml_entries:
```

Comment:
I think this loads all the stuff into memory at once? Using iterators and processing files one by one is probably the way to go? I.e. `glob.iglob` to read file names and then `with open()` inside the for-loop to read contents to close file handles.